### ISSUE TITLE
Hide bump seeds in all PDA accounts

### ISSUE THREAD

**Issue by armaniferrante** [solana-foundation/anchor#1039]
We can completely hide bump seeds for all PDA accounts *with data* by introducing a single byte to the account header, holding the bump, where the bump is accessible via the `Bump` trait.

**Comment by callensm** [comment]
by "header" are you referring to the account discriminator, which would change it from 8 bytes to 9?

**Comment by armaniferrante** [comment]
> by "header" are you referring to the account discriminator, which would change it from 8 bytes to 9 that must be pre-allocated for every account?

Yes and potential other fields in the future, e.g., a version number.

**Comment by callensm** [comment]
this would be a cool addition. unfortunately would end up being a permanently breaking change for all accounts created prior to `v1.0.0` correct?

**Comment by armaniferrante** [comment]
> this would be a cool addition. unfortunately would end up being a permanently breaking change for all accounts created prior to `v1.0.0` correct?

Correct.

**Comment by armaniferrante** [comment]
Proposal for the new 8 byte account header

```
<1-byte-anchor-layout-version> || <1-byte-account-version> || <4-byte-discriminator> || <1-byte-bump> || <1-byte-unused>
```

Additionally, a new IDL field would need to be added for the `anchor-layout-version`, so that apps know how to decode account data. If there is no such field, then we can assume the current and soon to be deprecated 8-byte discriminator is being used.

**Comment by paul-schaaf** [comment]
correct me if Im wrong
- 0: determines serialization algo used
- 1: changes if user changes the data structure e.g. adds a field
- [2-5]: same discriminator as before just 4 bytes
- 6: pda bump
- 7: reserved

**Comment by armaniferrante** [comment]
Correct. Though to clarify for others, 0 is used to version + define the the serialization + layout of the entire account--including the header.

**Comment by armaniferrante** [comment]
We actually don't need `0` because it can be stored in the IDL which is needed to unpack the account anyway.
